File: templates/Enterprise-Template/src/typescript/enterprise-bot/src/dialogs/main/mainResponses.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    ActionTypes,
    Activity,
    Attachment,
    CardFactory,
    InputHints,
    MessageFactory,
    TurnContext } from 'botbuilder';
import * as i18n from 'i18n';
import {
    DictionaryRenderer,
    LanguageTemplateDictionary,
    TemplateFunction } from '../templateManager/dictionaryRenderer';
import { TemplateManager } from '../templateManager/templateManager';

export class MainResponses extends TemplateManager {

    // Fields
    public static responseIds: {
        Cancelled: string;
        Completed: string;
        Confused: string;
        Greeting: string;
        Help: string;
        Intro: string;
    } = {
        Cancelled: 'cancelled',
        Completed: 'completed',
        Confused: 'confused',
        Greeting:  'greeting',
        Help:  'help',
        Intro:  'intro'
    };
    private static readonly responseTemplates: LanguageTemplateDictionary = new Map([
        ['default', new Map([
            [MainResponses.responseIds.Cancelled, MainResponses.fromResources('main.cancelled')],
            [MainResponses.responseIds.Completed, MainResponses.fromResources('main.completed')],
            [MainResponses.responseIds.Confused, MainResponses.fromResources('main.confused')],
            [MainResponses.responseIds.Greeting, MainResponses.fromResources('main.greeting')],
            [MainResponses.responseIds.Help,
                // tslint:disable-next-line:no-any
                (context: TurnContext, data: any): Promise<Activity> => MainResponses.buildHelpCard(context, data)],
            [MainResponses.responseIds.Intro,
                // tslint:disable-next-line:no-any
                (context: TurnContext, data: any): Promise<Activity> => MainResponses.buildIntroCard(context, data)]
        ])]
    ]);

    constructor() {
        super();
        this.register(new DictionaryRenderer(MainResponses.responseTemplates));
    }

    // tslint:disable-next-line:no-any
    public static buildIntroCard(turnContext: TurnContext, data: any): Promise<Activity> {
        const introPath: string = i18n.__('main.introPath');
        // tslint:disable-next-line:no-any non-literal-require
        const introCard: any = require(introPath);
        const attachment: Attachment = CardFactory.adaptiveCard(introCard);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, '', attachment.content.speak, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: i18n.__('main.helpBtnText1'),
                    type: ActionTypes.ImBack,
                    value: i18n.__('main.helpBtnValue1')
                },
                {
                    title: i18n.__('main.helpBtnText2'),
                    type: ActionTypes.ImBack,
                    value: i18n.__('main.helpBtnValue2')
                },
                {
                    title: i18n.__('main.helpBtnText3'),
                    type: ActionTypes.OpenUrl,
                    value: i18n.__('main.helpBtnValue3')
                }
            ],
            to: []
        };

        return Promise.resolve(<Activity> response);
    }

    // tslint:disable-next-line:no-any
    public static async buildHelpCard(turnContext: TurnContext, data: any): Promise<Activity> {
        const title: string = i18n.__('main.helpTitle');
        const text: string = i18n.__('main.helpText');
        const attachment: Attachment = CardFactory.heroCard(title, text);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, text, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
            {
                title: i18n.__('main.helpBtnText1'),
                type: ActionTypes.ImBack,
                value: i18n.__('main.helpBtnValue1')
            },
            {
                title: i18n.__('main.helpBtnText2'),
                type: ActionTypes.ImBack,
                value: i18n.__('main.helpBtnValue2')
            },
            {
                title: i18n.__('main.helpBtnText3'),
                type: ActionTypes.OpenUrl,
                value: i18n.__('main.helpBtnValue3')
            }
            ],
            to: []
        };

        return Promise.resolve(<Activity> response);
    }

    private static fromResources(name: string): TemplateFunction {
        return (): Promise<string> => Promise.resolve(i18n.__(name));
    }
}

File: samples/assistants/HospitalitySample/Responses/Main/MainResponses.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.IO;
using AdaptiveCards;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Builder.TemplateManager;
using Microsoft.Bot.Schema;

namespace HospitalitySample.Responses.Main
{
    public class MainResponses : TemplateManager
    {
        private static LanguageTemplateDictionary _responseTemplates = new LanguageTemplateDictionary
        {
            ["default"] = new TemplateIdMap
            {
                {
                    ResponseIds.Cancelled,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.CANCELLED,
                        ssml: MainStrings.CANCELLED,
                        inputHint: InputHints.AcceptingInput)
                },
                {
                    ResponseIds.Completed,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.COMPLETED,
                        ssml: MainStrings.COMPLETED,
                        inputHint: InputHints.AcceptingInput)
                },
                {
                    ResponseIds.Confused,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.CONFUSED,
                        ssml: MainStrings.CONFUSED,
                        inputHint: InputHints.AcceptingInput)
                },
                {
                    ResponseIds.Greeting,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.GREETING,
                        ssml: MainStrings.GREETING,
                        inputHint: InputHints.AcceptingInput)
                },
                { ResponseIds.Help, (context, data) => BuildHelpCard(context, data) },
                { ResponseIds.NewUserGreeting, (context, data) => BuildNewUserGreetingCard(context, data) },
                { ResponseIds.ReturningUserGreeting, (context, data) => BuildReturningUserGreetingCard(context, data) },
            }
        };

        public MainResponses()
        {
            Register(new DictionaryRenderer(_responseTemplates));
        }

        public static IMessageActivity BuildNewUserGreetingCard(ITurnContext turnContext, dynamic data)
        {
            var introCard = File.ReadAllText(MainStrings.INTRO_PATH);
            var card = AdaptiveCard.FromJson(introCard).Card;
            
            // work around for adaptive card actions not working in Teams
            if (!Channel.SupportsSuggestedActions(turnContext.Activity.ChannelId))
            {
                introCard += File.ReadAllText(MainStrings.INTRO_PATH_TEAMS);
                card = AdaptiveCard.FromJson(introCard).Card;
            }

            var attachment = new Attachment(AdaptiveCard.ContentType, content: card);

            var response = MessageFactory.Attachment(attachment, ssml: card.Speak, inputHint: InputHints.AcceptingInput);
            if (Channel.SupportsSuggestedActions(turnContext.Activity.ChannelId))
            {
                response.SuggestedActions = new SuggestedActions { Actions = GetCardActions() };
            }
            else
            {
                response.Attachments.Add(new HeroCard(buttons: GetCardActions()).ToAttachment());
            }

            return response;
        }

        public static IMessageActivity BuildReturningUserGreetingCard(ITurnContext turnContext, dynamic data)
        {
            var introCard = File.ReadAllText(MainStrings.INTRO_RETURNING);
            var card = AdaptiveCard.FromJson(introCard).Card;
            var attachment = new Attachment(AdaptiveCard.ContentType, content: card);

            var response = MessageFactory.Attachment(attachment, ssml: card.Speak, inputHint: InputHints.AcceptingInput);

            response.SuggestedActions = new SuggestedActions
            {
                Actions = GetCardActions()
            };

            return response;
        }

        public static IMessageActivity BuildHelpCard(ITurnContext turnContext, dynamic data)
        {
            var attachment = new HeroCard()
            {
                Title = MainStrings.HELP_TITLE,
                Text = MainStrings.HELP_TEXT,
            };

            var response = MessageFactory.Attachment(attachment.ToAttachment(), ssml: MainStrings.HELP_TEXT, inputHint: InputHints.AcceptingInput);

            if (Channel.SupportsSuggestedActions(turnContext.Activity.ChannelId))
            {
                response.SuggestedActions = new SuggestedActions { Actions = GetCardActions() };
            }
            else
            {
                response.Attachments.Add(new HeroCard(buttons: GetCardActions()).ToAttachment());
            }

            return response;
        }

        private static List<CardAction> GetCardActions()
        {
            return new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_1, value: MainStrings.HELP_BTN_VALUE_1),
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_2, value: MainStrings.HELP_BTN_VALUE_2),
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_3, value: MainStrings.HELP_BTN_VALUE_3),
                };
        }

        public class ResponseIds
        {
            // Constants
            public const string Cancelled = "cancelled";
            public const string Completed = "completed";
            public const string Confused = "confused";
            public const string Greeting = "greeting";
            public const string Help = "help";
            public const string NewUserGreeting = "newUser";
            public const string ReturningUserGreeting = "returningUser";
        }
    }
}
